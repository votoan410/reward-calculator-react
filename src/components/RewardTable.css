.table__header-green {
  background-color: green;
}

.table__header-yellow {
  background-color: yellow;
}
