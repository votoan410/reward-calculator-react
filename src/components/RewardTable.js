import { useState, useEffect } from "react";
import calculateRewardPoints from "../utils/pointsCalculate";

export default function RewardTable(props) {
  const [totalPoints, setTotalPoints] = useState({});

  const { transactionData, loading } = props;

  useEffect(() => {
    let userTotalPay = {};
    if (transactionData) {
      // map user to their respective total pay
      transactionData.forEach((transaction) => {
        // add user and current points to the new object
        if (!userTotalPay.hasOwnProperty(transaction.userId)) {
          userTotalPay[transaction.userId] = calculateRewardPoints(
            transaction.value
          );
        } 
        // if the user already exists within the new object, then just accumulate the point
        else {
          userTotalPay[transaction.userId] += calculateRewardPoints(
            transaction.value
          );
        }
      });
    }
    setTotalPoints(userTotalPay);
  }, [transactionData]);

  return (
    <>
      {loading ? (
        <div>Table is waiting for data...</div>
      ) : (
        <>
          <table data-testid="table">
            <thead>
              <tr>
                <th>user Id</th>
                <th> Transaction Date</th>
                <th className="table__header-green"> Payment Value $$$</th>
                <th className="table__header-yellow"> Calculated Rewards</th>
              </tr>
            </thead>
            <tbody>
              {transactionData.map((transaction, idx) => (
                // use index as key in this case because that is the only stable and unique data
                // for the current application
                <tr key={idx}>
                  <td>{transaction.userId}</td>
                  <td>
                    {new Date(transaction.date).toDateString()}{" "}
                    {new Date(transaction.date).toTimeString()}
                  </td>
                  <td>${transaction.value}</td>
                  <td>{calculateRewardPoints(transaction.value)}</td>
                </tr>
              ))}
            </tbody>
          </table>
          <table data-testid="table">
            <thead>
              <tr>
                <th>User Id</th>
                <th className="table__header-yellow">
                  Total Calculated Reward Points
                </th>
              </tr>
            </thead>
            <tbody>
              {/* the USER ID will be used as key in here since it will be unique in this case */}
              {Object.keys(totalPoints).map((entry) => (
                <tr key={entry}>
                  <td>{entry}</td>
                  <td>{totalPoints[entry]}</td>
                </tr>
              ))}
            </tbody>
          </table>
        </>
      )}
    </>
  );
}
