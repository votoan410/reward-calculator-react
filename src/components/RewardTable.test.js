import { render } from "@testing-library/react";
import RewardTable from "./RewardTable";

const mockDataTable = [
  {
    value: 50,
    date: "2021-01-11T06:22:49",
    userId: 1,
  },
  {
    value: 10,
    date: "2021-01-08T22:00:30",
    userId: 1,
  },

  {
    value: 100,
    date: "2021-01-09T22:28:41",
    userId: 3,
  },
  {
    value: 120,
    date: "2021-01-07T01:49:13",
    userId: 3,
  },
  {
    value: 100,
    date: "2021-01-15T14:34:20",
    userId: 2,
  },

  {
    value: 500,
    date: "2021-02-17T11:47:47",
    userId: 2,
  },
  {
    value: 36,
    date: "2021-01-02T18:57:37",
    userId: 1,
  },
  {
    value: 49,
    date: "2021-02-19T15:37:58",
    userId: 2,
  },
  {
    value: 35,
    date: "2021-03-27T14:29:15",
    userId: 2,
  },
  {
    value: 50,
    date: "2021-03-21T10:37:09",
    userId: 3,
  },
  {
    value: 45,
    date: "2021-03-11T07:02:40",
    userId: 3,
  },
  {
    value: 99,
    date: "2021-03-04T22:43:01",
    userId: 1,
  },
];

describe("Reward Table", () => {
  it("Should match the snapshot", () => {
    const { asFragment } = render(
      <RewardTable loading={false} transactionData={mockDataTable} />
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render the table", () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const wrappedComponent = render(
      <RewardTable loading={false} transactionData={mockDataTable} />
    );
    expect(wrappedComponent).not.toBeNull();
    // expect to render 2 tables: 1 contains transactions and respective reward points, 1 contains the total points calculated
    // eslint-disable-next-line testing-library/prefer-screen-queries
    expect(wrappedComponent.queryAllByTestId("table").length).toEqual(2);
  });

  it("Shouldn't render the table when loading is true (i.e, transactions data is not yet ready)", () => {
    // eslint-disable-next-line testing-library/render-result-naming-convention
    const { getByText } = render(
      <RewardTable loading={true} transactionData={mockDataTable} />
    );
    // eslint-disable-next-line testing-library/prefer-screen-queries
    const loadingText = getByText("Table is waiting for data...");
    expect(loadingText).toBeTruthy();
  });
});
