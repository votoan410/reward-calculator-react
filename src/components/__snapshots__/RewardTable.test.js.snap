// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reward Table Should match the snapshot 1`] = `
<DocumentFragment>
  <table
    data-testid="table"
  >
    <thead>
      <tr>
        <th>
          user Id
        </th>
        <th>
           Transaction Date
        </th>
        <th
          class="table__header-green"
        >
           Payment Value $$$
        </th>
        <th
          class="table__header-yellow"
        >
           Calculated Rewards
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          Mon Jan 11 2021 06:22:49 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $50
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          1
        </td>
        <td>
          Fri Jan 08 2021 22:00:30 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $10
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Sat Jan 09 2021 22:28:41 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $100
        </td>
        <td>
          50
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Thu Jan 07 2021 01:49:13 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $120
        </td>
        <td>
          90
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          Fri Jan 15 2021 14:34:20 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $100
        </td>
        <td>
          50
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          Wed Feb 17 2021 11:47:47 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $500
        </td>
        <td>
          850
        </td>
      </tr>
      <tr>
        <td>
          1
        </td>
        <td>
          Sat Jan 02 2021 18:57:37 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $36
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          Fri Feb 19 2021 15:37:58 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $49
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          Sat Mar 27 2021 14:29:15 GMT-0500 (Central Daylight Time)
        </td>
        <td>
          $35
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Sun Mar 21 2021 10:37:09 GMT-0500 (Central Daylight Time)
        </td>
        <td>
          $50
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          Thu Mar 11 2021 07:02:40 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $45
        </td>
        <td>
          0
        </td>
      </tr>
      <tr>
        <td>
          1
        </td>
        <td>
          Thu Mar 04 2021 22:43:01 GMT-0600 (Central Standard Time)
        </td>
        <td>
          $99
        </td>
        <td>
          49
        </td>
      </tr>
    </tbody>
  </table>
  <table
    data-testid="table"
  >
    <thead>
      <tr>
        <th>
          User Id
        </th>
        <th
          class="table__header-yellow"
        >
          Total Calculated Reward Points
        </th>
      </tr>
    </thead>
    <tbody>
      <tr>
        <td>
          1
        </td>
        <td>
          49
        </td>
      </tr>
      <tr>
        <td>
          2
        </td>
        <td>
          900
        </td>
      </tr>
      <tr>
        <td>
          3
        </td>
        <td>
          140
        </td>
      </tr>
    </tbody>
  </table>
</DocumentFragment>
`;
